lib/router/src/cache.rs

```diff
@@ -122,14 +113,20 @@ where
     ///
     /// Polls the underlying `DelayQueue`. When elements are returned from the
     /// queue, remove the associated key from `values`.
-    fn poll_purge(&mut self) -> Poll<(), Error> {
-        while let Some(expired) = try_ready!(self.expirations.poll()) {
-            debug!("expiring an item from the cache");
-            self.values.remove(expired.get_ref());
+    fn poll_purge(&mut self) -> Poll<(), ()> {
+        loop {
+            match try_ready!(self
+                .expirations
+                .poll()
+                .map_err(|e| unreachable!("expiration must not fail: {}", e)))
+            {
+                None => return Ok(Async::NotReady),
```

Comment:
What are the semantics of expirations returning None here?  Does that mean that there will never be any expirations ever again?  By my reading, returning NotReady here causes PurgeCache.poll to return NotReady but doesn't schedule this task to ever get polled again?

Reply:
Great question! From the DelayQueue docs:

> If no items are in the queue, i.e. is_empty() returns true, then poll returns Ready(None). This indicates that the stream has reached an end. However, if a new item is inserted after, poll will once again start returning items or `NotReady`.

But the real question is whether the delay queue notifies in this case...

FWIW, this seems to work fine in my manual tests of this behavior, but may not be correct by contract... We'd want to poll on something that will be notified when an item is inserted to be sure.